src/lib/storage/index/partial_hash/partial_hash_index_impl.cpp

```diff
@@ -116,86 +118,81 @@ PartialHashIndexImpl<DataType>::PartialHashIndexImpl(
+  return indexed_chunks;
 }
 
 template <typename DataType>
 size_t PartialHashIndexImpl<DataType>::remove_entries(const std::vector<ChunkID>& chunks_to_unindex) {
-  const auto size_before = _indexed_chunk_ids.size();
+  const size_t size_before = _indexed_chunk_ids.size();
 
   auto indexed_chunks_to_unindex = std::unordered_set<ChunkID>{};
   for (const auto& chunk_id : chunks_to_unindex) {
-    if (!_indexed_chunk_ids.contains(chunk_id)) {
+    if (!_indexed_chunk_ids.contains(chunk_id))
       continue;
-    }
 
     indexed_chunks_to_unindex.insert(chunk_id);
     _indexed_chunk_ids.erase(chunk_id);
   }
 
   // Checks whether a given RowID's ChunkID is in the set of ChunkIDs to be unindexed.
   auto is_to_unindex = [&indexed_chunks_to_unindex](const RowID& row_id) {
-    return indexed_chunks_to_unindex.contains(row_id.chunk_id);
+    const auto find_iter =
+        std::find(indexed_chunks_to_unindex.cbegin(), indexed_chunks_to_unindex.cend(), row_id.chunk_id);
+    return find_iter != indexed_chunks_to_unindex.cend();
   };
 
   // Iterate over all values stored in the index.
   auto map_iter = _map.begin();
-  auto keys_to_delete = std::unordered_set<DataType>{};
-
   while (map_iter != _map.end()) {
-    typename tbb::concurrent_hash_map<DataType, std::vector<RowID>>::accessor hash_map_accessor;
-    _map.find(hash_map_accessor, map_iter->first);
-    auto& row_ids = hash_map_accessor->second;
+    auto& row_ids = _map.at(map_iter->first);
 
     // Remove every RowID entry of the value that references one of the chunks.
     row_ids.erase(std::remove_if(row_ids.begin(), row_ids.end(), is_to_unindex), row_ids.end());
 
-    if (row_ids.empty()) {
-      keys_to_delete.insert(hash_map_accessor->first);
-    }
-    ++map_iter;
+    map_iter = row_ids.empty() ? _map.erase(map_iter) : ++map_iter;
   }
 
   auto& nulls = _null_values;
-  nulls.erase(std::remove_if(nulls.begin(), nulls.end(), is_to_unindex), nulls.end());
-
-  for (const auto& key_to_delete : keys_to_delete) {
-    _map.erase(key_to_delete);
-  }
 
+  nulls.erase(std::remove_if(nulls.begin(), nulls.end(), is_to_unindex), nulls.end());
 
   return size_before - _indexed_chunk_ids.size();
 }
 
 template <typename DataType>
 typename PartialHashIndexImpl<DataType>::IteratorPair PartialHashIndexImpl<DataType>::range_equals(
     const AllTypeVariant& value) const {
-  auto range = _map.equal_range(boost::get<DataType>(value));
-  return std::make_pair(Iterator(std::make_shared<TableIndexFlattenedSparseMapIterator<DataType>>(range.first)),
-                        Iterator(std::make_shared<TableIndexFlattenedSparseMapIterator<DataType>>(range.second)));
+  const auto begin = _map.find(boost::get<DataType>(value));
+  if (begin == _map.end()) {
+    const auto end_iter = cend();
+    return std::make_pair(end_iter, end_iter);
+  }
+  auto end = begin;
+  return std::make_pair(Iterator(std::make_shared<TableIndexSparseMapIterator<DataType>>(begin)),
```

Comment:
You create an Iterator by passing a shared_ptr to an iterator? Is that necessary?

Reply:
Ah okay, there are a lot of `using` along the way. Could you check if it would be possible to call `Iterator(begin)` or maybe even `Iterator.wrap(begin)` which internally creates the shared pointer? We don't do anything here with the shared pointer and having a `wrap` method would make that entire idea clearer.

Reply:
I'm unsure if this is possible, since the Wrapper defined inside `AbstractTableIndex` doesn't know about ´TableIndexIterator` and ´TableIndexNullValuesIterator`.
I was able to put the pointer stuff inside the ÌteratorWrapper`, but the acquired Iterators now behave like `BaseTableIndexIterator`s (e.g. Failing when trying to dereference them), which renders them useless.

Reply:
Here is an approach to improve readability: https://github.com/hyrise/hyrise/pull/2448/commits/b090fd29f5b0831b842d25ef726681872001288b